File: rust/crates/hyphae-testsupport/src/fake_cli.rs
//! The faked `claude` seam every client leaf drives: recorded envelopes in, calls out.
//!
//! The twin of `tests/enrich/fake_cli.py`. No process starts here. [`FakeCli`] implements
//! `CliRunner`, answers from the recorded envelopes in `tests/enrich/fixtures/` — read by path,
//! so both tiers pin the same recording — and records every call, which is what makes the argv,
//! the constructed env, the temp cwd and the deadline assertable.
//!
//! Only two answer envelopes were recorded, a success and the logged-out error, plus the two
//! auth blobs. Every other shape is a mutation of the success, labelled where it is built.

use std::collections::{BTreeMap, HashMap, HashSet};
use std::sync::{Arc, Condvar, Mutex};
use std::time::{Duration, Instant};

use hyphae_enrich::Answer;
use hyphae_enrich::client::{EnrichRequest, ITEM_TIMEOUT};
use hyphae_enrich::runner::{Call, CallError, CliRunner, Output};
use hyphae_enrich::validation::FailureKind;
use serde_json::{Map, Value};

use crate::corpus::repo;

/// The model the fake answers are attributed to, at both doors that write rows.
pub const MODEL: &str = "claude-haiku-4-5-20251001";

/// A second model id, for the leaves where agreeing with `DEFAULT_MODEL` by accident would
/// hide a client that ignored the model it was built with.
pub const OTHER_MODEL: &str = "claude-sonnet-4-5-20250929";

/// A short stand-in for the real per-level instructions: the client forwards whatever it is
/// given, so the text only has to be recognizable in an argv assertion.
pub const INSTRUCTIONS: &str = "Describe the item you are about to read.";

/// What the fake calls the `claude auth status` call, which passes no input.
pub const AUTH_CALL: &str = "<auth status>";

/// How long a gated fake waits for the call before it in the completion chain. A ceiling, not
/// a pace: a mistake in the chain fails the test instead of hanging the run.
pub const GATE_TIMEOUT: Duration = Duration::from_secs(10);

/// One recorded envelope, fresh each call so a mutation cannot leak between leaves.
///
/// # Panics
/// When the fixture is missing or is not JSON.
pub fn recorded(name: &str) -> Value {
    let path = repo()
        .join("tests/enrich/fixtures")
        .join(format!("{name}.json"));
    let text = std::fs::read_to_string(&path).expect("the recorded envelope is readable");
    serde_json::from_str(&text).expect("the recorded envelope is JSON")
}

/// The recorded success envelope with fields replaced — a derived shape, not a recording.
pub fn mutated(changes: &[(&str, Value)]) -> Value {
    let mut envelope = recorded("envelope_success");
    let fields = envelope.as_object_mut().expect("the envelope is an object");
    for (field, value) in changes {
        fields.insert((*field).to_owned(), value.clone());
    }
    envelope
}

/// The recorded success envelope with fields removed — derived, standing for CLI drift.
pub fn without(fields: &[&str]) -> Value {
    let mut envelope = recorded("envelope_success");
    let held = envelope.as_object_mut().expect("the envelope is an object");
    for field in fields {
        held.remove(*field).expect("the envelope carried the field");
    }
    envelope
}

/// The structured output the recorded success carries, which is what the client hands back.
pub fn recorded_output() -> Map<String, Value> {
    recorded("envelope_success")["structured_output"]
        .as_object()
        .expect("the recorded answer is an object")
        .clone()
}

/// The recorded call's own usage numbers, re-keyed by a caller to stand for another model.
pub fn recorded_usage() -> Value {
    recorded("envelope_success")["modelUsage"][MODEL].clone()
}

/// What the item with this key renders to. The fake keys its script on this.
pub fn content_of(key: &str) -> String {
    format!("# Main turn\n\nrender for {key}")
}

/// One request per key, all carrying the same stand-in instructions.
pub fn requests_for(keys: &[&str]) -> Vec<EnrichRequest> {
    keys.iter()
        .map(|key| EnrichRequest {
            key: (*key).to_owned(),
            instructions: INSTRUCTIONS.to_owned(),
            content: content_of(key),
        })
        .collect()
}

/// Every answer by key: its failure kind, or None where the model answered.
pub fn kinds(answers: &[Answer]) -> BTreeMap<String, Option<FailureKind>> {
    answers
        .iter()
        .map(|answer| (answer.key().to_owned(), answer.failure()))
        .collect()
}

/// One scripted `claude` call: what it writes, or what it fails with instead.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Reply {
    pub stdout: String,
    pub code: i32,
    /// A process that never answered, as the runner reports one.
    pub fails: Option<CallError>,
}

impl Reply {
    /// A call that exited zero with this on stdout.
    pub fn printing(stdout: &str) -> Self {
        Self {
            stdout: stdout.to_owned(),
            code: 0,
            fails: None,
        }
    }

    /// A call that exited nonzero, printing whatever it printed.
    pub fn refusing(stdout: &str) -> Self {
        Self {
            stdout: stdout.to_owned(),
            code: 1,
            fails: None,
        }
    }

    /// A call that produced no process at all.
    pub fn failing(error: CallError) -> Self {
        Self {
            stdout: String::new(),
            code: 0,
            fails: Some(error),
        }
    }

    /// Whether this reply carries a usable envelope — what ends the serial canary phase.
    fn answers(&self) -> bool {
        if self.fails.is_some() || self.code != 0 {
            return false;
        }
        // Stdout the client cannot read is no more an answer than an errored call is.
        serde_json::from_str::<Value>(&self.stdout)
            .is_ok_and(|envelope| envelope.get("is_error") == Some(&Value::Bool(false)))
    }
}

/// The recorded success: an answer, as it really came back on 2026-08-13.
pub fn succeeds() -> Reply {
    Reply::printing(&recorded("envelope_success").to_string())
}

/// The recorded logged-out call: exit 1, `is_error`, no answer.
pub fn errors() -> Reply {
    Reply::refusing(&recorded("envelope_logged_out").to_string())
}

/// A hung process, killed at the item deadline.
pub fn hangs() -> Reply {
    Reply::failing(CallError::Timeout(ITEM_TIMEOUT))
}

/// Stands in for a real process: answers from a script, and records every call.
///
/// Scripted by item content, which is unique per key, so a retry of the same item gets the same
/// reply. `gate` optionally holds a call open until another call has started, which is how a
/// completion order that is not the submission order gets forced.
pub struct FakeCli {
    replies: HashMap<String, Reply>,
    gate: Option<Gate>,
    seen: Mutex<Seen>,
}

/// What holds a call open until the chain releases it.
type Gate = Box<dyn Fn(&str) + Send + Sync>;

/// What the fake watched happen, behind one lock.
#[derive(Debug, Default)]
struct Seen {
    calls: Vec<Call>,
    started: Vec<String>,
    live: usize,
    answered: usize,
    peak_before_an_answer: usize,
}

impl FakeCli {
    /// A fake answering these replies, keyed by the content each call sends.
    pub fn new(replies: HashMap<String, Reply>) -> Arc<Self> {
        Arc::new(Self {
            replies,
            gate: None,
            seen: Mutex::new(Seen::default()),
        })
    }

    /// The same, with every call held at `gate` before it is answered.
    pub fn gated(
        replies: HashMap<String, Reply>,
        gate: impl Fn(&str) + Send + Sync + 'static,
    ) -> Arc<Self> {
        Arc::new(Self {
            replies,
            gate: Some(Box::new(gate)),
            seen: Mutex::new(Seen::default()),
        })
    }

    /// Every call the fake was asked for, whole, in start order.
    pub fn calls(&self) -> Vec<Call> {
        self.seen.lock().expect("the fake's lock").calls.clone()
    }

    /// Every content the fake was asked for, in call-start order.
    pub fn started(&self) -> Vec<String> {
        self.seen.lock().expect("the fake's lock").started.clone()
    }

    /// The widest overlap seen while no call had yet returned an answer — 1 means the calls
    /// before the first answer were strictly serial.
    pub fn peak_before_an_answer(&self) -> usize {
        self.seen
            .lock()
            .expect("the fake's lock")
            .peak_before_an_answer
    }
}

impl CliRunner for FakeCli {
    fn run(&self, call: &Call) -> Result<Output, CallError> {
        let key = call.input.clone().unwrap_or_else(|| AUTH_CALL.to_owned());
        {
            let mut seen = self.seen.lock().expect("the fake's lock");
            seen.calls.push(call.clone());
            seen.started.push(key.clone());
            seen.live += 1;
            if seen.answered == 0 {
                seen.peak_before_an_answer = seen.peak_before_an_answer.max(seen.live);
            }
        }
        if let Some(gate) = &self.gate {
            gate(&key);
        }
        let reply = self
            .replies
            .get(&key)
            .unwrap_or_else(|| panic!("the fake has no reply scripted for {key:?}"))
            .clone();
        {
            let mut seen = self.seen.lock().expect("the fake's lock");
            seen.live -= 1;
            if reply.answers() {
                seen.answered += 1;
            }
        }
        match reply.fails {
            Some(error) => Err(error),
            None => Ok(Output {
                code: reply.code,
                stdout: reply.stdout,
            }),
        }
    }
}

/// Forces a completion order by holding each call until another call has started.
///
/// A call starting is the only thing a fake can see that proves the client recorded an earlier
/// answer: the client feeds one new item per answer it records. So each gated call waits for
/// the *start* of the item fed by the completion it must follow.
#[derive(Debug)]
pub struct Chain {
    waits_for: HashMap<String, String>,
    started: Mutex<HashSet<String>>,
    woken: Condvar,
}

impl Chain {
    /// A chain in which each key waits for the key it maps to.
    pub fn new(waits_for: HashMap<String, String>) -> Arc<Self> {
        Arc::new(Self {
            waits_for,
            started: Mutex::new(HashSet::new()),
            woken: Condvar::new(),
        })
    }

    /// Hold this call until the call it waits for has started.
    ///
    /// # Panics
    /// When the awaited call never starts, so a mistake in the chain fails a leaf instead of
    /// hanging the run.
    pub fn gate(&self, key: &str) {
        let mut started = self.started.lock().expect("the chain's lock");
        started.insert(key.to_owned());
        self.woken.notify_all();
        let Some(awaited) = self.waits_for.get(key) else {
            return;
        };
        let deadline = Instant::now() + GATE_TIMEOUT;
        while !started.contains(awaited) {
            let left = deadline.saturating_duration_since(Instant::now());
            assert!(!left.is_zero(), "{key:?} waited for {awaited:?} to start");
            started = self
                .woken
                .wait_timeout(started, left)
                .expect("the chain's lock")
                .0;
        }
    }
}

File: tests/enrich/fake_cli.py
"""The faked `claude` seam every client test drives: recorded envelopes in, invocations out.

No process starts here. `subprocess.run` is replaced by `FakeCli`, which answers from the
recorded envelopes in `fixtures/` and records every invocation — so the argv, the constructed
env, the temp cwd and the deadline are all assertable. The seam is the module attribute the
client really resolves, not a private helper, and the guard in `test_no_live_api.py` is what
proves that: an unpatched client trips it rather than launching `claude`.

Only two envelopes are recorded: a success and the logged-out error. Every other shape is a
mutation of the success, labelled where it is built. A plain module rather than a conftest,
so a reader of either client test file can see the whole fake in one place.
"""

import json
import subprocess
import threading
from collections.abc import Callable, Mapping, Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import pytest

from hyphae.enrich.client import (
    CLAUDE,
    ITEM_TIMEOUT,
    EnrichRequest,
    Failed,
    Result,
)
from hyphae.enrich.validation import FailureKind

FIXTURES = Path(__file__).parent / "fixtures"


MODEL = "claude-haiku-4-5-20251001"


# A second model id, for the tests where agreeing with `DEFAULT_MODEL` by accident would hide
# a client that ignored the model it was built with.
OTHER_MODEL = "claude-sonnet-4-5-20250929"


# Short stand-ins for the real per-level instructions: the client forwards whatever it is
# given, so the text only has to be recognizable in an argv assertion.
INSTRUCTIONS = "Describe the item you are about to read."


# How long a gated fake waits for the call before it in the completion chain. A ceiling, not
# a pace: a mistake in the chain fails the test instead of hanging the run.
GATE_TIMEOUT = 10.0


def recorded(name: str) -> dict[str, Any]:
    """One recorded envelope, fresh each call so a mutation cannot leak between tests."""
    return json.loads((FIXTURES / f"{name}.json").read_text())


def mutated(**changes: Any) -> dict[str, Any]:
    """The recorded success envelope with fields replaced — a derived shape, not a recording."""
    return recorded("envelope_success") | changes


def without(*fields: str) -> dict[str, Any]:
    """The recorded success envelope with fields removed — derived, standing for CLI drift."""
    envelope = recorded("envelope_success")
    for name in fields:
        del envelope[name]
    return envelope


# The recorded call's own usage numbers, re-keyed below to stand for a substituted model.
RECORDED_USAGE = recorded("envelope_success")["modelUsage"][MODEL]


def content_of(key: str) -> str:
    """What the item with this key renders to. The fake keys its script on this."""
    return f"# Main turn\n\nrender for {key}"


def requests_for(*keys: str) -> list[EnrichRequest]:
    return [
        EnrichRequest(key=key, instructions=INSTRUCTIONS, content=content_of(key)) for key in keys
    ]


@dataclass(frozen=True)
class Reply:
    """One scripted `claude` invocation: what it writes, or what it raises instead."""

    stdout: str = ""
    # What the CLI wrote on its diagnostic stream, which is the only stream a failure keeps.
    stderr: str = ""
    returncode: int = 0
    # A hung process, as `subprocess.run` reports one.
    raises: BaseException | None = None

    @property
    def answers(self) -> bool:
        """Whether this reply carries a usable envelope — what ends the serial canary phase."""
        if self.raises is not None or self.returncode != 0:
            return False
        try:
            envelope = json.loads(self.stdout)
        except json.JSONDecodeError:
            # Stdout the client cannot read is no more an answer than an errored call is.
            return False
        return not envelope.get("is_error", True)


def succeeds() -> Reply:
    return Reply(stdout=json.dumps(recorded("envelope_success")))


def errors() -> Reply:
    """The recorded logged-out call: exit 1, `is_error`, no answer."""
    return Reply(stdout=json.dumps(recorded("envelope_logged_out")), returncode=1)


def refused(stdout: str = "") -> Reply:
    """The recorded refusal: exit 1, one line on stderr, and nothing on stdout.

    What `claude` writes when it is passed a flag it does not take — the shape a CLI version
    bump takes, and the one that fails every item of a round identically. `stdout` plants an
    answer beside it (invented), for the leaves that prove no failure quotes that stream.
    """
    return Reply(
        stdout=stdout,
        stderr=(FIXTURES / "stderr_unknown_option.txt").read_text(),
        returncode=1,
    )


def hangs() -> Reply:
    return Reply(raises=subprocess.TimeoutExpired(CLAUDE, ITEM_TIMEOUT))


class FakeCli:
    """Stands in for `subprocess.run`: answers from a script, and records every call.

    Scripted by item content, which is unique per key, so a retry of the same item gets the
    same reply. `gate` optionally holds a call open until another call has started, which is
    how a completion order that is not the submission order gets forced.
    """

    def __init__(
        self,
        replies: Mapping[str, Reply],
        *,
        gate: Callable[[str], None] | None = None,
    ) -> None:
        self.replies = replies
        self.gate = gate
        self.calls: list[dict[str, Any]] = []
        # Every content the fake was asked for, in call-start order.
        self.started: list[str] = []
        self._lock = threading.Lock()
        self._live = 0
        self._answered = 0
        # The widest overlap seen while no call had yet returned an answer — 1 means the
        # calls before the first answer were strictly serial.
        self.peak_before_an_answer = 0

    def __call__(self, argv: Sequence[str], **kwargs: Any) -> subprocess.CompletedProcess[str]:
        key = kwargs.get("input", AUTH_CALL)
        with self._lock:
            self.calls.append({"argv": list(argv), **kwargs})
            self.started.append(key)
            self._live += 1
            if not self._answered:
                self.peak_before_an_answer = max(self.peak_before_an_answer, self._live)
        try:
            if self.gate is not None:
                self.gate(key)
            reply = self.replies[key]
            if reply.raises is not None:
                raise reply.raises
            return subprocess.CompletedProcess(
                list(argv), reply.returncode, reply.stdout, reply.stderr
            )
        finally:
            with self._lock:
                self._live -= 1
                if key in self.replies and self.replies[key].answers:
                    self._answered += 1

    def install(self, monkeypatch: pytest.MonkeyPatch) -> "FakeCli":
        monkeypatch.setattr(subprocess, "run", self)
        return self


# What the fake calls the `claude auth status` call, which passes no `input`.
AUTH_CALL = "<auth status>"


class Chain:
    """Forces a completion order by holding each call until another call has started.

    A call starting is the only thing a fake can see that proves the client recorded an
    earlier result: the client feeds one new item per completion it records. So each gated
    call waits for the *start* of the item fed by the completion it must follow.
    """

    def __init__(self, waits_for: Mapping[str, str]) -> None:
        self.waits_for = waits_for
        self.started: dict[str, threading.Event] = {}
        self._lock = threading.Lock()

    def _event(self, key: str) -> threading.Event:
        with self._lock:
            return self.started.setdefault(key, threading.Event())

    def __call__(self, key: str) -> None:
        self._event(key).set()
        awaited = self.waits_for.get(key)
        if awaited is not None and not self._event(awaited).wait(timeout=GATE_TIMEOUT):
            raise RuntimeError(f"{key!r} waited {GATE_TIMEOUT}s for {awaited!r} to start")


Install = Callable[..., FakeCli]


def kinds(results: Sequence[Result]) -> dict[str, FailureKind | None]:
    """Every result by key: its failure kind, or None where the model answered."""
    return {result.key: result.kind if isinstance(result, Failed) else None for result in results}
